src/backend/d3d12/CommandBufferD3D12.cpp

```diff
@@ -44,8 +187,59 @@ namespace d3d12 {
     }
 
     void CommandBuffer::FillCommands(ComPtr<ID3D12GraphicsCommandList> commandList) {
+        BindGroupStateTracker bindingTracker(device);
```

Comment:
Done. A large CPU only heap is created and then copied into a smaller GPU-visible heap of the right size.